api/common/sharing/utils.go

```diff
@@ -10,56 +12,105 @@ import (
 	"github.com/Peripli/service-manager/pkg/web"
 	"github.com/Peripli/service-manager/storage"
 	"github.com/tidwall/gjson"
+	"reflect"
 )
 
-func ExtractReferencedInstanceID(ctx context.Context, repository storage.Repository, body []byte, tenantIdentifier string, getTenantId func() string) (string, error) {
+func ExtractReferencedInstanceID(req *web.Request, repository storage.Repository, body []byte, tenantIdentifier string, getTenantId func() string, smaap bool) (string, error) {
+	var err error
+	ctx := req.Context()
 	parameters := gjson.GetBytes(body, "parameters").Map()
-	referencedInstanceID, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
 
-	if !exists {
-		return "", util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	err = validateParameters(parameters)
+	if err != nil {
+		return "", err
+	}
+
+	sharedInstancesMap := map[string]*types.ServiceInstance{}
+	sharedInstancesMap, err = getSharedInstancesByTenant(ctx, repository, tenantIdentifier, getTenantId())
+	if err != nil {
+		return "", err
 	}
 
+	var referencedInstance *types.ServiceInstance
+	referencedInstanceID := parameters[instance_sharing.ReferencedInstanceIDKey].String()
+	if len(referencedInstanceID) > 1 {
+		referencedInstance, err = filterInstancesByID(ctx, repository, referencedInstanceID, tenantIdentifier, getTenantId())
+		if err != nil {
+			return "", err
+		}
+	} else {
+		filteredInstancesMap := map[string]*types.ServiceInstance{}
+		filteredInstancesMap, err = filterInstancesBySelectors(ctx, repository, sharedInstancesMap, parameters, smaap)
+		err = validateSelectorResults(filteredInstancesMap, parameters)
+		if err != nil {
+			return "", err
+		}
+
+		referencedInstance, err = retrieveInstanceFromMap(sharedInstancesMap)
```

Comment:
what this function does???
your looping over map and return the last result

Reply:
as discussed: moved from a map to a list of instances.
removed the converters functions